workflows/prognostic_c48_run/runtime/loop.py

```diff
@@ -106,32 +111,86 @@ def fillna_tendency(tendency: xr.DataArray) -> Tuple[xr.DataArray, xr.DataArray]
+    dQv = dQv.assign_attrs(units="m/s/s").astype(np.float64, casting="same_kind")
+    return dQu, dQv
+
+
+def transform_from_agrid_to_dgrid(
+    u: xr.DataArray, v: xr.DataArray
+) -> Tuple[xr.DataArray, xr.DataArray]:
+    """Transform a vector field on the A-grid in latitude-longitude coordinates
+    to the D-grid in cubed-sphere coordinates.
+
+    u and v must have double precision and contain like units attributes.
+    """
+    u_quantity = pace.util.Quantity.from_data_array(u)
+    v_quantity = pace.util.Quantity.from_data_array(v)
+    (
+        x_wind_quantity,
+        y_wind_quantity,
+    ) = fv3gfs.wrapper.transform_agrid_winds_to_dgrid_winds(u_quantity, v_quantity)
+    return x_wind_quantity.data_array, y_wind_quantity.data_array
+
+
+def add_tendency(state: Any, tendencies: State, dt: float) -> Tuple[State, State]:
     """Given state and tendency prediction, return updated state, which only includes
     variables updated by tendencies. Also returns column-integrated fraction of
     tendencies which are filled nans.
     """
 
     with xr.set_options(keep_attrs=True):
         updated = {}
-        tendency_filled_frac = {}
-        for name_ in tendency:
-            name = str(name_)
-            try:
-                state_name = str(TENDENCY_TO_STATE_NAME[name])
-            except KeyError:
-                raise KeyError(
-                    f"Tendency variable '{name}' does not have an entry mapping it "
-                    "to a corresponding state variable to add to. "
-                    "Existing tendencies with mappings to state are "
-                    f"{list(TENDENCY_TO_STATE_NAME.keys())}"
-                )
-            (
-                tendency_filled,
-                tendency_filled_frac[f"{name}_filled_frac"],
-            ) = fillna_tendency(tendency[name])
-            updated[state_name] = state[state_name] + tendency_filled * dt
-    return updated, tendency_filled_frac  # type: ignore
+        filled_tendencies = {}
+        filled_fractions = {}
+
+        for name, tendency in tendencies.items():
+            filled, filled_fraction = fillna_tendency(tendency)
+            filled_tendencies[name] = filled
+            filled_fractions[f"{name}_filled_frac"] = filled_fraction
+
+        # First update x_wind and y_wind based on the A-grid wind tendencies. If
+        # D-grid wind tendencies are provided later, update x_wind and y_wind
+        # within the for loop.  Tendency diagnostics for dQu, dQv, dQx_wind, and
+        # dQy_wind will be recorded separately.
+        if set(filled_tendencies).intersection(A_GRID_WIND_TENDENCIES):
+            dQu, dQv = prepare_agrid_wind_tendencies(filled_tendencies)
+            transformed = transform_from_agrid_to_dgrid(dQu, dQv)
+            dQx_wind_from_agrid, dQy_wind_from_agrid = transformed
+            updated[X_WIND] = state[X_WIND] + dQx_wind_from_agrid * dt
+            updated[Y_WIND] = state[Y_WIND] + dQy_wind_from_agrid * dt
```

Comment:
Maybe better to just not implement case of dQu and dQx_wind tendencies? Feels like a bit of a red flag that there are now multiple `updated = state + tendency * dt` lines in this function.

What about just having this `add_tendency` function assuming all provided tendencies are on the same grid as the corresponding state variable? i.e. do the A->D grid transformation (if necessary) in separate preceding call?

Reply:
Sure thing -- I agree there's not really a strong reason to continue supporting that.  

I'll refactor the na-filling out of `add_tendency` in that case as well, since it needs to occur prior to the coordinate transformation.  Along similar lines there's maybe an argument that it should have been that way all along.